### ISSUE TITLE
Unable to run multiple BatchSchedulers

### ISSUE THREAD

**Issue by imranrajjad** [EasyBatch/easybatch-framework#46]
I am calling the below code snippet inside a loop 

BatchScheduler scheduler = new BatchScheduler(engine);	        
Date startDate = new Date(new Date().getTime() + 1);
scheduler.scheduleAtWithInterval(startDate, intervalInMinutes);
scheduler.start();

--

basically I am trying to schedule multiple engines , each with its own reader,mapper and processor.

stack trace is as below:

org.easybatch.integration.quartz.BatchSchedulerException: An exception occurred during scheduler startup
	at org.easybatch.integration.quartz.BatchScheduler.start(BatchScheduler.java:124)
	at com.lmkt.we.job.JobManager.startDataPull(JobManager.java:64)
	at com.lmkt.we.bootstrap.BootStrap.contextInitialized(BootStrap.java:35)
	at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4765)
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5260)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:150)
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1525)
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1515)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
	at java.util.concurrent.FutureTask.run(FutureTask.java:138)
	at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:895)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:918)
	at java.lang.Thread.run(Thread.java:662)
Caused by: org.quartz.ObjectAlreadyExistsException: Unable to store Job : 'DEFAULT.batch-job', because one already exists with this identification.
	at org.quartz.simpl.RAMJobStore.storeJob(RAMJobStore.java:279)
	at org.quartz.simpl.RAMJobStore.storeJobAndTrigger(RAMJobStore.java:251)
	at org.quartz.core.QuartzScheduler.scheduleJob(QuartzScheduler.java:886)
	at org.quartz.impl.StdScheduler.scheduleJob(StdScheduler.java:249)
	at org.easybatch.integration.quartz.BatchScheduler.start(BatchScheduler.java:119)
	... 12 more

**Comment by benas** [comment]
Hi Imran,

Thank you for pointing out this issue. In fact, scheduling multiple engines fails even if you use separate instances of the BatchScheduler. Here is how I reproduced the issue:

```java
Engine engine1 = new EngineBuilder().build();
Engine engine2 = new EngineBuilder().build();

BatchScheduler scheduler1 = new BatchScheduler(engine1);
scheduler1.scheduleAtWithInterval(new Date(), 1);
scheduler1.start();

BatchScheduler scheduler2 = new BatchScheduler(engine2);
scheduler2.scheduleAtWithInterval(new Date(), 1);
scheduler2.start();
```

I get the same exception:

```java
org.easybatch.integration.quartz.BatchSchedulerException: An exception occurred during scheduler startup
	at org.easybatch.integration.quartz.BatchScheduler.start(BatchScheduler.java:124)
	at org.easybatch.integration.quartz.BatchSchedulerTest.testJobScheduling(BatchSchedulerTest.java:31)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:77)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:195)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:63)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:120)
Caused by: org.quartz.ObjectAlreadyExistsException: Unable to store Job : 'DEFAULT.batch-job', because one already exists with this identification.
	at org.quartz.simpl.RAMJobStore.storeJob(RAMJobStore.java:279)
	at org.quartz.simpl.RAMJobStore.storeJobAndTrigger(RAMJobStore.java:251)
	at org.quartz.core.QuartzScheduler.scheduleJob(QuartzScheduler.java:886)
	at org.quartz.impl.StdScheduler.scheduleJob(StdScheduler.java:249)
	at org.easybatch.integration.quartz.BatchScheduler.start(BatchScheduler.java:119)
	... 27 more
```

This is due to the use of the same in-memory `SchedulerFactory` with the same (default) job name.

I'll fix that asap in version 3.0.1-SNAPSHOT and I'll keep you informed once the fix is deployed to maven central.

Cheers
Mahmoud

**Comment by imranrajjad** [comment]
yes that is exactly how I it occurred, the workaround is to implement your own RecordReader and use single scheduler for the multiple jobs.

Also please note that if this code is to be called inside a J2EE webapp, the container starts complaining about a thread that has not finished and maycause memory leak
-----
SEVERE: The web application [/we] appears to have started a thread named [DefaultQuartzScheduler_QuartzSchedulerThread] but has failed to stop it. This is very likely to create a memory leak.
.

**Comment by imranrajjad** [comment]
yes that is exactly how I it occurred, the workaround is to implement your own RecordReader and use single scheduler for the multiple jobs.

Also please note that if this code is to be called inside a J2EE webapp, the container starts complaining about a thread that has not finished and maycause memory leak
-----
SEVERE: The web application [/we] appears to have started a thread named [DefaultQuartzScheduler_QuartzSchedulerThread] but has failed to stop it. This is very likely to create a memory leak.
.

**Comment by imranrajjad** [comment]
yes that is exactly how I it occurred, the workaround is to implement your own RecordReader and use single scheduler for the multiple jobs.

Also please note that if this code is to be called inside a J2EE webapp, the container starts complaining about a thread that has not finished and maycause memory leak
-----
SEVERE: The web application [/we] appears to have started a thread named [DefaultQuartzScheduler_QuartzSchedulerThread] but has failed to stop it. This is very likely to create a memory leak.
.

**Comment by benas** [comment]
As you may have read in the [job scheduling tutorial](http://www.easybatch.org/tutorials/quartz.html), the `BatchScheduler` API is a wrapper around the [Quartz](http://quartz-scheduler.org/)'s low level APIs. so behind the scene, it's quartz  that is creating threads and easy batch has no control over them.

If you are using Easy Batch in a JEE server, you **should** use the thread executor service of your server and **not rely** on the BatchScheduler API. The engine is actually a `java.util.concurrent.Callable` so it can be scheduled using a `java.util.concurrent.ScheduledExecutorService` in a Java SE environment, or a `javax.enterprise.concurrent.ManagedExecutorService` in Java EE environment.

That said, I am not sure I understand your workaround using a custom record reader. A possible solution I am thinking about is to give every engine instance a unique Id, and use this id as the job Id for the quartz scheduler (This will prevent creating a scheduler with an existing id).

Last question, why has your thread failed to finish? Can you give me more details please?

**Comment by imranrajjad** [comment]
sorry for late reply, I agree on ManagedExecutorService  in JavaEE, but that would make the application container dependent, for instance, the web app should be deployable on Jetty and tomcat also, ManagedExecutorService requires a container like JBoss or Weblogic etc.

About the custom record reader, I implemented the open() method in such a way that it handled all the jobs, eventually creating a single record,which is actually an array of records.

**Comment by benas** [comment]
Hi Imran,

I have deployed version 3.0.1-SNAPSHOT in maven central repository with the fix for this issue. Every job has now a unique Id which allow scheduling multiple engines ([This test](https://github.com/EasyBatch/easybatch-framework/blob/master/easybatch-integration/easybatch-quartz/src/test/java/org/easybatch/integration/quartz/BatchSchedulerTest.java#L52) is passing now).

Moreover, I have added a method to stop the scheduler (`BatchScheduler.stop()`) which will allow you to control when to start/stop each scheduler. You can also check the status of the scheduler with
the new `isStarted` and `isStopped` methods.

Could you please retest with this new version and keep me informed if it is ok for you?

Many thanks upfront
Mahmoud

**Comment by benas** [comment]
Hello Imran,

FYI, this fix has been released in version 3.0.1.

Regards
Mahmoud
